packages/streamr-logger/include/streamr-logger/Logger.hpp

```diff
@@ -1,18 +1,152 @@
-#ifndef STREAMER_UTILS_LOGGER_LOGGER_HPP
-#define STREAMER_UTILS_LOGGER_LOGGER_HPP
+#ifndef STREAMER_UTILS_LOGGER_LOGGER_H
+#define STREAMER_UTILS_LOGGER_LOGGER_H
 
-#include <iostream>
+#include <folly/logging/LogConfig.h>
+#include <folly/logging/LogLevel.h>
+#include <folly/logging/LogMessage.h>
+#include <folly/logging/LogWriter.h>
+#include <folly/logging/LoggerDB.h>
+#include <folly/logging/xlog.h>
+#include "StreamrWriterFactory.hpp"
+#include "StreamrHandlerFactory.hpp"
+
+namespace streamr {
+namespace logger {
+
+namespace {
+
+static const auto envLogLevelName = "LOG_LEVEL";
+static const auto logLevelTraceText = "trace";
+static const auto logLevelDebugText = "debug";
+static const auto logLevelInfoText = "info";
+static const auto logLevelWarnText = "warn";
+static const auto logLevelErrorText = "error";
+static const auto logLevelFatalText = "fatal";
+
+enum class StreamrLogLevel { trace, debug, info, warn, error, fatal };
+
+const std::unordered_map<StreamrLogLevel, folly::LogLevel> fromStreamrToFollyLogLevelMap{
+    {StreamrLogLevel::trace, folly::LogLevel::DBG},
+    {StreamrLogLevel::debug, folly::LogLevel::DBG0},
+    {StreamrLogLevel::info, folly::LogLevel::INFO},
+    {StreamrLogLevel::warn, folly::LogLevel::WARN},
+    {StreamrLogLevel::error, folly::LogLevel::ERR},
+    {StreamrLogLevel::fatal, folly::LogLevel::CRITICAL}};
+
+const std::unordered_map<std::string, folly::LogLevel> toFollyLogLevelMap{
+    {logLevelTraceText, folly::LogLevel::DBG},
+    {logLevelDebugText, folly::LogLevel::DBG0},
+    {logLevelInfoText, folly::LogLevel::INFO},
+    {logLevelWarnText, folly::LogLevel::WARN},
+    {logLevelErrorText, folly::LogLevel::ERR},
+    {logLevelFatalText, folly::LogLevel::CRITICAL}};
+
+std::optional<folly::LogLevel> getFollyLogLevelFromEnv() {
+    char* val = getenv(envLogLevelName);
+    if (val) {
+        auto pair = toFollyLogLevelMap.find(val);
+        if (pair != toFollyLogLevelMap.end()) {
+            return std::optional<folly::LogLevel>{std::move(pair->second)};
+        }
+    }
+    return std::nullopt;
+}
+
+}; // namespace
 
-namespace streamr::logger {
 class Logger {
 public:
-    Logger() = default;
 
-    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
-    void log(const std::string& message) const {
-        std::cout << message << std::endl;
+    Logger(std::shared_ptr<folly::LogWriter> logWriter = nullptr)
+        : loggerDB_{folly::LoggerDB::get()} {
+        if (logWriter) {
+            writerFactory_ = StreamrWriterFactory(logWriter);
+        } else {
+            writerFactory_ = StreamrWriterFactory();
+        }
+        logHandlerFactory_ = {
+            std::make_unique<StreamrHandlerFactory>(&writerFactory_)};
+        this->initializeLoggerDB(folly::LogLevel::DBG);
+    }
+
+    static Logger& get(std::shared_ptr<folly::LogWriter> logWriter = nullptr) {
```

Comment:
Is this class meant to be singleton or why is this needed? How many copies of the class can you create? When are they deleted? Is there a risk for memory leaks?

Reply:
It was the idea to use it the class without an explicit initialization. We discussed it earlier. So yes it was a singleton. So no really risk for memory leaks and not even data race because it was Meyer's Singleton. Anyway I removed it if it is not needed.